telemetry/internal/appinsights/client.go

```diff
@@ -0,0 +1,184 @@
+		batchInterval := cmp.Or(c.MaxBatchInterval, 10*time.Second)
+		httpClient := cmp.Or(c.HTTPClient, http.DefaultClient)
+		c.channel = newInMemoryChannel(endpoint, batchSize, batchInterval, httpClient)
+		c.context = setupContext(c.InstrumentationKey, c.Tags)
+		if err := contracts.SanitizeTags(c.context.Tags); err != nil {
+			c.ErrorLog.Printf("Warning sanitizing tags: %v", err)
+		}
+		c.initialized.Store(true)
+	})
+}
+
+func setupContext(instrumentationKey string, tags map[string]string) *telemetryContext {
+	context := newTelemetryContext(instrumentationKey)
+	context.Tags["ai.internal.sdkVersion"] = sdkName + ":" + version
+	maps.Copy(context.Tags, tags)
+	return context
+}
+
+// Enabled returns true if this client is enabled and will accept telemetry.
+func (c *Client) Enabled() bool {
+	return !c.disabled
+}
+
+// SetEnabled enables or disables the client. When disabled, telemetry is
+// silently swallowed by the client. Defaults to enabled.
+func (c *Client) SetEnabled(enabled bool) {
+	c.disabled = !enabled
+}
+
+// NewEvent creates a new event with the specified name.
+func (c *Client) NewEvent(name string) *Event {
+	return &Event{
+		name:   name,
+		client: c,
+	}
+}
+
+// TrackEvent logs a user action with the specified name.
+func (c *Client) TrackEvent(name string) {
+	c.NewEvent(name).Inc()
+}
+
+// Forces the current queue to be sent.
+func (c *Client) Flush() {
+	if !c.initialized.Load() {
+		return
+	}
+	c.channel.flush()
+}
+
+// Close flushes and tears down the submission goroutine and closes internal channels.
+// Waits until all pending telemetry items have been submitted.
+func (c *Client) Close(ctx context.Context) error {
+	if !c.initialized.Load() {
+		return nil
+	}
+	var d time.Duration
+	if timeout, ok := ctx.Deadline(); ok {
+		d = time.Until(timeout)
+		if d == 0 {
+			// Don't confuse the user with a zero duration.
+			d = -1
+		}
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-c.channel.close(d):
+		// TODO: check for errors.
+		return nil
+	}
+}
+
+// Tears down the submission goroutines, closes internal channels.
+// Any telemetry waiting to be sent is discarded.
+// This is a more abrupt version of [Client.Close].
+func (c *Client) Stop() {
+	if !c.initialized.Load() {
+		return
+	}
+	c.channel.stop()
+}
+
+// Submits the specified telemetry item.
+func (c *Client) track(data contracts.EventData) {
+	if !c.disabled {
+		c.init()
+		ev := c.context.envelop(data)
+		if err := ev.Sanitize(); err != nil {
+			c.ErrorLog.Printf("Warning sanitizing telemetry item: %v", err)
+		}
+		c.channel.send(c.context.envelop(data))
+	}
+}
+
+// Event represents an event to be tracked.
+type Event struct {
+	name   string
+	client *Client
+}
+
+// Inc adds 1 to the counter.
+func (e *Event) Inc() {
+	e.Add(1)
+}
+
+// Add adds n to the counter. n cannot be negative, as counts cannot decrease.
+func (e *Event) Add(n uint64) {
+	if n == 0 {
+		return
+	}
+	for range n {
+		e.client.track(contracts.EventData{
```

Comment:
It feels like there should be some practical limit here because of the wasteful-seeming way they're queued up to be transmitted... maybe at least some caution in the doc comment.

Reply:
I've optimized this method a bit. I still have to add tests for this method to see if how well it behave.